store/tikv/txn_scheduler.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv/latch"
+	binlog "github.com/pingcap/tipb/go-binlog"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type txnCommitter struct {
+	committer *twoPhaseCommitter
+	ch        chan error
+	ctx       context.Context
+	lock      *latch.Lock
+}
+
+func newTxnCommiter(ctx context.Context, committer *twoPhaseCommitter) *txnCommitter {
+	return &txnCommitter{
+		committer: committer,
+		ctx:       ctx,
+	}
+}
+
+func (txn *txnCommitter) commit(timeout bool) uint64 {
+	if timeout {
+		err := errors.Errorf("startTS %d timeout", txn.committer.startTS)
+		log.Debug(txn.committer.connID, err)
+		txn.ch <- errors.Annotate(err, txnRetryableMark)
+		return uint64(0)
+	}
+
+	commitTS, err := txn.execute()
+	txn.ch <- errors.Trace(err)
+	log.Debug(txn.committer.connID, " finish txn with startTS:", txn.committer.startTS, " commitTS:", commitTS, " error:", err)
+	return commitTS
+}
+
+func (txn *txnCommitter) execute() (commitTS uint64, err error) {
+	err = txn.committer.execute(txn.ctx)
+	if err != nil {
+		txn.committer.writeFinishBinlog(binlog.BinlogType_Rollback, 0)
+	} else {
+		commitTS = txn.committer.commitTS
+		txn.committer.writeFinishBinlog(binlog.BinlogType_Commit, int64(commitTS))
+	}
+	return
+}
+
+type txnMap struct {
+	txns map[uint64]*txnCommitter
+	sync.RWMutex
+}
+
+func (t *txnMap) put(startTS uint64, txn *txnCommitter) {
+	t.Lock()
+	defer t.Unlock()
+	if t.txns == nil {
+		t.txns = make(map[uint64]*txnCommitter)
+	}
+	if _, ok := t.txns[startTS]; ok {
+		panic(fmt.Sprintf("The startTS %d shouldn't be used in two transaction", txn.committer.startTS))
```

Comment:
Don't panic here, it will crash the tidb-server.

Reply:
I think it's a disaster if two or more transactions share the same `startTS`, I think panic should be ok here.